src/managers/backup.py

```diff
@@ -5,34 +5,43 @@
 """Helpers for managing backups."""
 
 import logging
+from datetime import datetime
+from io import StringIO
 
 import boto3
+import httpx
+import yaml
 from botocore import loaders, regions
 from botocore.exceptions import ClientError
 from mypy_boto3_s3.service_resource import Bucket
+from rich.console import Console
+from rich.table import Table
 
-from core.stubs import S3ConnectionInfo
+from core.cluster import ClusterState
+from core.stubs import BackupMetadata, S3ConnectionInfo
+from literals import ADMIN_SERVER_PORT
 
 logger = logging.getLogger(__name__)
 
 
 class BackupManager:
     """Manager for all things backup-related."""
 
-    def __init__(self, s3_parameters: S3ConnectionInfo) -> None:
-        self.s3_parameters = s3_parameters
+    def __init__(self, state: ClusterState) -> None:
+        self.state = state
 
     @property
     def bucket(self) -> Bucket:
         """S3 bucket to read from and write to."""
+        s3_parameters = self.state.cluster.s3_credentials
         s3 = boto3.resource(
             "s3",
-            aws_access_key_id=self.s3_parameters["access-key"],
-            aws_secret_access_key=self.s3_parameters["secret-key"],
-            region_name=self.s3_parameters["region"] if self.s3_parameters["region"] else None,
-            endpoint_url=self._construct_endpoint(self.s3_parameters),
+            aws_access_key_id=s3_parameters["access-key"],
+            aws_secret_access_key=s3_parameters["secret-key"],
```

Comment:
**thought/nonblocking:** I think I might be weird/wrong in this regard, but I'm allergic to `raise`-ing.
I tend to prefer not doing direct key `get`s that might raise `KeyError` and bork the charm, instead preferring to default to something safe, and then doing something like:

```python
value = dict.get("key", "")
if not value:
    return  # i.e don't execute something
```

I'm not suggesting that you should change that here, it's totally OK to keep. I just wonder if I'm 'right' or 'wrong' for thinking that way.

Reply:
I usually do the same, so I would be on the "you're right" side. In this specific case, we know for a fact that the values are present because the only execution branch leading to this code requires this dict to be validated in `_on_s3_credentials_changed` (and then stored in the peer cluster relation).
Would you like to have some kind of comment here to avoid having to follow the full execution logic?

Reply:
Nah not needed, was just a thought of mine generally. I'll resolve this.

Reply:
My opinion on this. If we plan on handling the key not existing, `.get()` is fine. Generally using get to me is like doing a try:catch all. If the key doesn't exist when we are sure it should exist, letting it raise is the better way address it.